lib/rethinkdb/connection.ex

```diff
@@ -123,213 +119,101 @@ defmodule RethinkDB.Connection do
-    end
-    state = %{
-      pending: %{},
-      current: {:start, ""},
-      token: 0,
-      config: Map.put(opts, :transport, {transport, transport_opts})
-    }
-    case sync_connect do
-      true ->
-        case connect(:sync, state) do
-          {:backoff, _, _} -> {:stop, :econnrefused}
-          x -> x
-        end
-      false ->
-        {:connect, :init, state}
-    end
-  end
-
-  def connect(_info, state = %{config: %{host: host, port: port, auth_key: auth_key, transport: {transport, transport_opts}}}) do
-    case Transport.connect(transport, host, port, [active: false, mode: :binary] ++ transport_opts) do
-      {:ok, socket} ->
-        case handshake(socket, auth_key) do
-          {:error, _} -> {:stop, :bad_handshake, state}
-          :ok ->
-            :ok = Transport.setopts(socket, [active: :once])
-            # TODO: investigate timeout vs hibernate
-            {:ok, Dict.put(state, :socket, socket)}
-        end
-      {:error, :econnrefused} ->
-        backoff = min(Dict.get(state, :timeout, 1000), 64000)
-        {:backoff, backoff, Dict.put(state, :timeout, backoff*2)}
-    end
-  end
-
-  def disconnect(info, state = %{pending: pending}) do
-    pending |> Enum.each(fn {_token, pid} ->
-      Connection.reply(pid, %RethinkDB.Exception.ConnectionClosed{})
-    end)
-    new_state = state
-      |> Map.delete(:socket)
-      |> Map.put(:pending, %{})
-      |> Map.put(:current, {:start, ""})
-    # TODO: should we reconnect?
-    {:stop, info, new_state}
-  end
-
-  def handle_call(:conn_opts, _from, state = %{config: opts}) do
-    {:reply, opts, state}
+  def noreply_wait(conn, timeout \\ 15_000) do
+    DBConnection.execute!(conn, %Q{message: "[3]"}, [], timeout: timeout, pool: __pool__(conn))
   end
 
-  def handle_call(_, _,
-    state = %{pending: pending, config: %{max_pending: max_pending}}) when map_size(pending) > max_pending do
-    {:reply, %RethinkDB.Exception.TooManyRequests{}, state}
-  end
-
-  def handle_call({:query_noreply, query}, _from, state = %{token: token}) do
-    new_token = token + 1
-    token = << token :: little-size(64) >>
-    {:noreply, state} = Request.make_request(query, token, :noreply, %{state | token: new_token})
-    {:reply, :noreply, state}
-  end
-
-  def handle_call({:query, query}, from, state = %{token: token}) do
-    new_token = token + 1
-    token = << token :: little-size(64) >>
-    Request.make_request(query, token, from, %{state | token: new_token})
-  end
-
-  def handle_call({:continue, token}, from, state) do
-    query = "[2]"
-    Request.make_request(query, token, from, state)
-  end
+  #
+  # DBConnection API
+  #
 
-  def handle_call({:stop, token}, from, state) do
-    query = "[3]"
-    Request.make_request(query, token, from, state)
+  def connect(options) do
+    # Starts the multiplexer process and stores its pid and some default options to the state.
+    {:ok, pid} = Multiplexer.start_link(options)
```

Comment:
I found that splitting up the `DBConnection` behaviour and the `Multiplexer` into two separate processes (with their own state and mailbox) could be a good thing.

The main reason is that `Multiplexer.send_recv/3` blocks until a reply is received (using `{:noreply, state}` and `GenServer.reply/1`on the server side).